### PR TITLE
FEATURE: add Optional countdown for e.g. deadlines

### PR DESCRIPTION
**Pull request by merefield**
Adds an optional "deadline" option to event setup:

image

That presents a countdown on the Topic List (and Topic):

image

* Also converts Event Form Component to `gjs`
* Adds some tests

### CODE REVIEW

File: config/locales/client.en.yml [hunk 1]
```diff
@@ -19,6 +22,19 @@ en:
       going_max: "Maximum guests"
       going_max_label: "{{goingMax}} Guests"
       error: "Event end should be same as or after event start"
+    event_label:
```
Comment (angusmcleod) [inline config/locales/client.en.yml:25]: Is there a reason why we're not using the existing day / hour / minute localisation in discourse/discourse? If we use that we also automatically get all of the translations, so there is a strong default to use that if the case is the same as it seems to be here.

https://github.com/discourse/discourse/blob/main/config/locales/server.en.yml#L943

File: assets/javascripts/discourse/lib/date-utilities.js.es6 [hunk 2]
```diff
@@ -361,10 +365,45 @@ function eventLabel(event, args = {}) {
         }
       }
     }
+
+    pastDue = moment() > start;
+
+    if (siteSettings.events_support_deadlines && event.deadline) {
+      deadline = true;
```
Comment (angusmcleod) [inline assets/javascripts/discourse/lib/date-utilities.js.es6:372]: There has to be a component in discourse that does this kind of display, for example for use in the countdown in a topic timer. Are you sure there isn't an existing helper function in `discourse/lib/formatter.js` that does this?

**Review by angusmcleod — CHANGES_REQUESTED** [review]
@merefield Generally looks ok, however I'm still not convinced there isn't something in Discourse we could be using for the countdown display itself. See my comments.

File: assets/javascripts/discourse/components/event-form.gjs [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+  @tracked goingMax;
+  @tracked usersGoing;
+
+  constructor() {
+    super(...arguments);
+    this.setupProperties();
+  }
+
+  @action
+  setupProperties() {
```
[the remark this answers is not in the document]
Comment (merefield) [reply assets/javascripts/discourse/components/event-form.gjs:43]: You are right, it's unnecessary, removed.

File: config/settings.yml [hunk 4]
```diff
@@ -87,3 +87,13 @@ plugins:
   events_add_default_end_time:
     default: false
     client: true
+  events_support_deadlines:
```
[the remark this answers is not in the document]
Comment (merefield) [reply config/settings.yml]: renamed

File: db/migrate/20240905012439_add_deadline_to_discourse_events_events.rb
[the remark this answers is not in the document]
Comment (merefield) [reply db/migrate/20240905012439_add_deadline_to_discourse_events_events.rb:1]: I thought you were migrating off of custom fields, so was trying to assist with that :)

I'll remove.

[the remark this answers is not in the document]
Comment (merefield) [reply db/migrate/20240905012439_add_deadline_to_discourse_events_events.rb:1]: Removed

**Comment by angusmcleod** [comment]
@merefield The issue was that you had added deadline as an attribute to the basic event serializer, however it was no longer an attribute on the model.

https://github.com/angusmcleod/discourse-events/commit/10246da59b7448d236ed7a90806a28dccb37d16d

**Comment by angusmcleod** [comment]
@merefield I took the opportunity to review the UI, and I think it still needs a bit more tweaking, in addition to anything arising out of my question concerning the use of core components.

Screenshot 2024-09-18 at 19 00 56

- It's a little cramped in the display in the composer, particularly if there's a timezone, i.e. the deadline icon is just a bit too close.
- It there's 0 hours the hours should not be displayed at all.

**Comment by merefield** [comment]
> @merefield The issue was that you had added deadline as an attribute to the basic event serializer, however it was no longer an attribute on the model.
> 
> [10246da](https://github.com/angusmcleod/discourse-events/commit/10246da59b7448d236ed7a90806a28dccb37d16d)

@angusmcleod it's still completely broken

File: config/locales/client.en.yml (hunk 1 above)
Comment (merefield) [reply config/locales/client.en.yml:25]: good catch!!

resolved and removed bespoke versions.

File: assets/javascripts/discourse/lib/date-utilities.js.es6 (hunk 2 above)
Comment (merefield) [reply assets/javascripts/discourse/lib/date-utilities.js.es6:372]: Unfortunately duration.humanize used for topic timers removes the detail by approximation which is not what we want.

Similar issue with https://github.com/discourse/discourse/blob/4bfe78d1e66fab345dda56f791e59c749d0de3a6/app/assets/javascripts/discourse/app/lib/formatter.js#L59

However, I've cleaned up the formatting removing commas and units when value is 0, e.g. no hours here so removed:

image

**Comment by angusmcleod** [comment]
@merefield I'm seeing a new exception when changing the timezone when editing an event. See this video:

https://www.loom.com/share/a2b9987eccf249b59b2603cd053ebbec?sid=714ec1ef-b013-4302-88b2-5b426d048713

**Comment by merefield** [comment]
> @merefield I'm seeing a new exception when changing the timezone when editing an event. See this video:
> 
> https://www.loom.com/share/a2b9987eccf249b59b2603cd053ebbec?sid=714ec1ef-b013-4302-88b2-5b426d048713

thanks Angus, I'll take a look

**Comment by merefield** [comment]
@angusmcleod that should be fixed now, apologies.

**Comment by merefield** [comment]
should be resolved

**Comment by angusmcleod** [comment]
@merefield Thanks. Just doing some more checking I see that the events with deadlines in the calendar popovers need some style improvements. The addition of the deadline squishes the content too much.

Screenshot 2024-09-19 at 15 06 44

**Comment by merefield** [comment]
> @merefield Thanks. Just doing some more checking I see that the events with deadlines in the calendar popovers need some style improvements. The addition of the deadline squishes the content too much.
> 
> Screenshot 2024-09-19 at 15 06 44
> Once that's addressed I'll merge this.

I'm wondering if I simply don't show them in this view?  Would that be acceptable?

**Comment by angusmcleod** [comment]
@merefield I'd be ok with that. As long as you and your user(s) is.

**Comment by merefield** [comment]
Improved:

**Comment by angusmcleod** [comment]
@merefield Thanks. Approved.

**Comment by merefield** [comment]
thanks for the review and merge @angusmcleod !

**Pull request merged**